src/mjolnir/traffic_updater.cc

```diff
@@ -0,0 +1,390 @@
+      }
+      if (speed.coefficients) {
+        tile_builder.AddPredictedSpeed(j, *speed.coefficients, pred_count);
+        directededge.set_has_predicted_speed(true);
+      }
+      ++stat.updated_count;
+    }
+
+    // Add the directed edge to the local list
+    directededges.emplace_back(std::move(directededge));
+  }
+
+  // Write the new tile with updated directed edges and the predicted speeds
+  tile_builder.UpdatePredictedSpeeds(directededges);
+}
+/**
+ * Read both the constrained and freeflow speed CSV files
+ * We expect the files to be named as <quadtreeID>.constrained.csv and
+ * <quadtreeID>.freeflow.csv. (e.g., 1202021.constrained.csv and 1202021.freeflow.csv)
+ */
+void UpdateTiles(const std::string& tile_dir,
+                 std::vector<std::pair<GraphId, std::vector<std::string>>>::const_iterator tile_start,
+                 std::vector<std::pair<GraphId, std::vector<std::string>>>::const_iterator tile_end,
+                 std::promise<TrafficStats>& result) {
+
+  std::stringstream thread_name;
+  thread_name << std::this_thread::get_id();
+
+  // Iterate through the tiles and parse them
+  size_t total = tile_end - tile_start;
+  double count = 0;
+  TrafficStats stat{};
+  for (; tile_start != tile_end; ++tile_start) {
+    LOG_INFO(thread_name.str() + " parsing traffic data for " + std::to_string(tile_start->first));
+    auto traffic = ParseTrafficFile(tile_start->second, stat);
+    LOG_INFO(thread_name.str() + " add traffic data to " + std::to_string(tile_start->first));
+    UpdateTile(tile_dir, tile_start->first, traffic, stat);
+    LOG_INFO(thread_name.str() + " finished " + std::to_string(tile_start->first) + "(" +
+             std::to_string(++count / total * 100.0) + ")");
+  }
+
+  result.set_value(stat);
+}
+std::vector<std::pair<GraphId, std::vector<std::string>>>
+PrepareTrafficTiles(const filesystem::path& traffic_tile_dir) {
+  std::unordered_map<GraphId, std::vector<std::string>> files_per_tile;
+  for (filesystem::recursive_directory_iterator i(traffic_tile_dir), end; i != end; ++i) {
+    if (i->is_regular_file()) {
+      // remove any extension
+      auto file_name = i->path().string();
+      auto pos = file_name.rfind(filesystem::path::preferred_separator);
+      file_name = file_name.substr(0, file_name.find('.', pos == std::string::npos ? 0 : pos));
+      try {
+        // parse it into a tile id and store the file path with it
+        auto id = GraphTile::GetTileId(file_name);
+        files_per_tile[id].push_back(i->path().string());
+      } catch (...) {}
+    }
+  }
+
+  std::vector<std::pair<GraphId, std::vector<std::string>>> traffic_tiles(files_per_tile.begin(),
+                                                                          files_per_tile.end());
+
+  std::random_device rd;
+  std::shuffle(traffic_tiles.begin(), traffic_tiles.end(), std::mt19937(rd()));
+
+  return traffic_tiles;
+}
+void LogProcessingResults(const TrafficStats& final_stats) {
```

Comment:
IMO there is no need to break this out into a function. Also, can you put an empty line between the function definitions?